src/main/java/com/nutomic/syncthingandroid/service/SyncthingRunnable.java

```diff
@@ -399,4 +384,49 @@ private void trimLogFile() {
         }
     }
 
+    private HashMap<String, String> buildEnvironment(SharedPreferences sp) {
+            HashMap<String, String> targetEnv = new HashMap<String, String>();
+            // Set home directory to data folder for web GUI folder picker.
+            targetEnv.put("HOME", Environment.getExternalStorageDirectory().getAbsolutePath());
+            targetEnv.put("STTRACE", sp.getString("sttrace", ""));
+            File externalFilesDir = mContext.getExternalFilesDir(null);
+            if (externalFilesDir != null)
+                targetEnv.put("STGUIASSETS", externalFilesDir.getAbsolutePath() + "/gui");
+            targetEnv.put("STNORESTART", "1");
+            targetEnv.put("STNOUPGRADE", "1");
+            // Disable hash benchmark for faster startup.
+            // https://github.com/syncthing/syncthing/issues/4348
+            targetEnv.put("STHASHING", "minio");
+            if (sp.getBoolean("use_tor", false)) {
+                targetEnv.put("all_proxy", "socks5://localhost:9050");
+                targetEnv.put("ALL_PROXY_NO_FALLBACK", "1");
+            }
+            if (sp.getBoolean("use_legacy_hashing", false))
+                targetEnv.put("STHASHING", "standard");
+            putCustomEnvironmentVariables(targetEnv, sp);
+            return targetEnv;
+    }
+
+    private Process setupAndLaunch(ProcessBuilder pb, HashMap<String, String> env) throws IOException {
+        Process process = null;
+        if (mUseRoot) {
+            process = pb.start();
+            // The su binary prohibits the inheritance of environment variables.
+            // Even with --preserve-environment the environment gets messed up.
+            // We therefore start a root shell, and set all the environment variables manually.
+            DataOutputStream suOut = new DataOutputStream(process.getOutputStream());
+            for (Map.Entry<String, String> entry : env.entrySet()) {
+                suOut.writeBytes(String.format("export %s=\"%s\"\n", entry.getKey(), entry.getValue()));
+            }
+            suOut.flush();
+            // Exec into Synchting, because we want to wait for Syncthing.
+            // Exec will replace the su process by Syncthing.
+            suOut.writeBytes("exec " + TextUtils.join(" ", mCommand) + "\n");
+            suOut = null;
```

Comment:
What's the reason for using exec here? I don't really understand from the comment.

Also, there's no reason to set suOut to null. But you should close() it.

Reply:
The exec function replaces the su process with Syncthing as execlp does in C.
If we do not use exec, we'll drop to the root shell as soon as Syncthing terminates, and we'll wait infinitely for the process to terminate (line 146).

I think when we close the stream, which is the stdin of su, su will terminate. This is the way I terminate the shell in fixAppDataPermissions as pointed out by @AudriusButkevicius .
We therefore should *not* close, but we want to get rid of the stream object.